lib/fastlane/plugin/wpmreleasetoolkit/actions/bundletool/bundletool_generate_universal_signed_apk_action.rb

```diff
@@ -0,0 +1,104 @@
+module Fastlane
+  module Actions
+    class BundletoolGenerateUniversalSignedApkAction < Action
+      def self.run(params)
+        aab_path = params[:aab_path]
+        apk_output_path = params[:apk_output_path]
+        keystore_path = params[:keystore_path]
+        keystore_password = params[:keystore_password]
+        keystore_key_alias = params[:keystore_key_alias]
+        signing_key_password = params[:signing_key_password]
+
+        begin
+          sh('command -v bundletool > /dev/null')
+        rescue StandardError
+          UI.user_error!('bundletool is not installed. Please install it using the instructions at https://developer.android.com/studio/command-line/bundletool')
+          raise
+        end
+
+        sh('bundletool', 'build-apks',
+           '--mode', 'universal',
+           '--bundle', aab_path,
+           '--output-format', 'DIRECTORY',
+           '--output', apk_output_path,
```

Comment:
I just tested the command, and the `--output` provided here is supposed to be a path to a directory, not the APK file.

For example, I tried the command on an `.aab` and passing `--output-format DIRECTORY --output foo/bar.apk` and it created the folders `foo` and `foo/bar.apk`, and generated a `foo/bar.apk/universal.apk` inside it.

We need more logic in this action to handle this
 - We should create a temp dir (`mktmpdir`) and make `bundletool` generate the out directory structure in this temp dir
 - Then, we move the `#{tmpdir}/universal.apk` file to the destination provided by the caller in `apk_output_path`
 - To go even further, if the `apk_output_path` is an existing `Dir.is_directory?`, or does not exist yet but the string ends with a `/`, then use `File.join(apk_output_path, File.basename(aab_path, '.aab) + '.apk')` as the destination to move the generated `universal.apk` to once `bundletool` is done. And otherwise, we consider that the `apk_output_path` is a path to a file and not a dir, and we use it directly as the destination to rename/`mv` the generated `.apk` to.

In any case, we should not keep the code as is because the `.apk` generated by `bundletool` would end up always be named `universal.apk` and be in a **subdirectory** provided by the `apk_output_path`, even if from a user of the action, it's likely that people will expect `apk_output_path` to be the path to the final `.apk`, and that will likely create a folder with extension `.apk` instead of getting the intended behavior.

Reply:
I guess the name `apk_output_path` might be a bit misleading in this context. I'll rename it to `apk_output_dir_path` instead to avoid confusion.

As you've already noted, the APK generated by `bundletool` will always be named `universal.apk`, so it's really just about the path to the directory where the APK will be generated. Maybe I should also mention it in the description of the parameter's `ConfigItem`.

Reply:
Given that we'll likely need to use the generated `.apk` (and thus need its path) in the following code of the lane calling that action (e.g. to upload it as artifact or upload it to Google Play or attach it to the GitHub Release, etc), we need to have this name be explicit at some point so that the code following this action knows what path to use to reference the generated file.

imho our options are:

1. Keep the path at which `bundletool` generates the `.apk` and the fact that it's always named `universal.apk`, and document it here in the action's `ConfigItem` description.
   - I'm not a fan of this because that means that if we wanted to e.g. rely on the `artifacts_paths: ['**/*.aab', '**/*.apk']` key in our pipeline to upload all AABs and APKs, the artefacts would appear as `app-base-release.aab` but `mydir/universal.apk`, making it inconsistent and also not clear what flavor and build type the APK was from.
   - Also that means we won't have a choice in the name of the generated APK
2. Make the `apk_output_path` be the path of the APK to be generated, not the path to the folder.
   - That simply means have `bundletool` use a temp dir for `--output #{tmpdir}`, then call `FileUtils.mv("#{tmpdir}/universal.apk", apk_output_path)` to move it to the expected destination path
   - Feels more flexible, and allows the caller to choose the exact path for the generated APK instead of having the file name forced on them.
3. Be even more flexible and detect if the `apk_output_path` is an existing folder, and if so use the same basename as the aab for the output apk
   - i.e. same as option 2, but just do something like `apk_output_path = File.join(apk_output_path, "#{File.basename(aab_path, '.aab')}.apk") if Dir.directory?(apk_output_path)` to adjust the `apk_output_path` before doing the final `FileUtils.mv`.

My preference would be for option 3, which is the most flexible all while not being that complicated to implement.

Reply:
Please see my comment [here](https://github.com/wordpress-mobile/release-toolkit/pull/467#discussion_r1184174852).  It deals with the `--output` option but uses a slightly different approach.

Reply:
I've now fixed the two security issues (path with spaces not working and leading to potential parameter injection, and risk of deleting an entire destination directory due to unchecked `rm -rf`) in 0e801e1be5b514e1b8c8c7717881991e916021a0 (+ added some unit tests in b28c1f02ecfa2ec49bc777ee072b52dbd0276219 to validate those issues were resolved)